src/dft/backends/mklcpu/mklcpu_helpers.hpp

```diff
@@ -0,0 +1,201 @@
+/*******************************************************************************
+* Copyright 2022 Codeplay Software Ltd.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions
+* and limitations under the License.
+*
+*
+* SPDX-License-Identifier: Apache-2.0
+*******************************************************************************/
+
+#ifndef _ONEMKL_DFT_SRC_MKLCPU_HELPERS_HPP_
+#define _ONEMKL_DFT_SRC_MKLCPU_HELPERS_HPP_
+
+#include "oneapi/mkl/detail/exceptions.hpp"
+#include "oneapi/mkl/dft/detail/types_impl.hpp"
+
+// MKLCPU header
+#include "mkl_dfti.h"
+
+namespace oneapi {
+namespace mkl {
+namespace dft {
+namespace mklcpu {
+namespace detail {
+
+template<typename T> struct is_buffer : std::false_type {};
+template<> struct is_buffer<sycl::buffer<float>> : std::true_type {};
+template<> struct is_buffer<sycl::buffer<std::complex<float>>> : std::true_type {};
+template<> struct is_buffer<sycl::buffer<double>> : std::true_type {};
+template<> struct is_buffer<sycl::buffer<std::complex<double>>> : std::true_type {};
+
+template<typename T> struct is_usm : std::false_type {};
+template<> struct is_usm<float> : std::true_type {};
+template<> struct is_usm<std::complex<float>> : std::true_type {};
+template<> struct is_usm<double> : std::true_type {};
+template<> struct is_usm<std::complex<double>> : std::true_type {};
+
+using mklcpu_desc_t = DFTI_DESCRIPTOR_HANDLE;
+using commit_t = dft::detail::commit_impl;
+template< class T>
+inline constexpr bool is_buffer_v = is_buffer<T>::value;
+template< class T>
+inline constexpr bool is_usm_v = is_usm<T>::value;
+
+// host_task automatically uses run_on_host_intel if it is supported by the
+//  compiler. Otherwise, it falls back to single_task.
+template <typename K, typename H, typename F>
+static inline auto host_task_internal(H &cgh, F f, int) -> decltype(cgh.host_task(f)) {
+    return cgh.host_task(f);
+}
+
+template <typename K, typename H, typename F>
+static inline void host_task(H &cgh, F f) {
+    (void)host_task_internal<K>(cgh, f, 0);
+}
```

Comment:
this is a derivation of the abstraction defined in [dfti_sycl.h](https://github.com/intel-innersource/libraries.performance.math.mkl/blob/f2b57101320809c81086c15374bc207be4c8b963/src/dft/include/dfti_sycl.h#L120), this might be useful (or was useful before we learnt there was no immidiate need to replace all the host_tasks by single_task) later in case we need to change this to single_task

Reply:
Unfortunately we don't access to that source code. Why would we need to replace `host_task` with `single_task`?

Reply:
I still don't understand this unfortunately. 

> this might be useful ... later in

I think the normal advice is YAGNI, and that we should add code when its needed.

Reply:
this is similary used in many other domains : 
https://github.com/oneapi-src/oneMKL/blob/a492e0acf01a9915f96659e84739e01754efe986/src/rng/backends/mklcpu/mrg32k3a.cpp#L152
https://github.com/oneapi-src/oneMKL/blob/a492e0acf01a9915f96659e84739e01754efe986/src/lapack/backends/rocsolver/rocsolver_task.hpp#L55
and many others